src/util/taproot.rs

```diff
@@ -451,19 +452,27 @@ impl TaprootBuilder {
 
     /// Creates a [`TaprootSpendInfo`] with the given internal key.
     ///
-    // TODO: in a future breaking API change, this no longer needs to return result
+    /// Returns the unmodified builder as Err if the builder is not finalizable.
+    /// See also [`TaprootBuilder::is_finalizable`]
     pub fn finalize<C: secp256k1::Verification>(
         mut self,
         secp: &Secp256k1<C>,
         internal_key: UntweakedPublicKey,
-    ) -> Result<TaprootSpendInfo, TaprootBuilderError> {
-        match self.branch.pop() {
-            None => Ok(TaprootSpendInfo::new_key_spend(secp, internal_key, None)),
-            Some(Some(node)) => {
-                Ok(TaprootSpendInfo::from_node_info(secp, internal_key, node))
+    ) -> Result<TaprootSpendInfo, TaprootBuilder> {
+        match self.branch.len() {
+            0 => Ok(TaprootSpendInfo::new_key_spend(secp, internal_key, None)),
+            1 => {
+                if let Some(Some(node)) = self.branch.pop() {
+                    Ok(TaprootSpendInfo::from_node_info(secp, internal_key, node))
+                } else {
+                    // This should be unreachable as
+                    // - The builder invariant guarantees that the last element is always Some
+                    // - The size of vector is checked to be 1
+                    // But still we return an error in order to avoid a panic reasoning.
+                    Err(self)
+                }
             }
-            _ => Err(TaprootBuilderError::IncompleteTree),
```

Comment:
Perhaps:


```rust
pub struct IncompleteTree {
    pub builder: TaprootBuilder,
    len: usize,
}

// impl Display & co
```

This would provide more information about the error while still allowing to keep the builder.

Reply:
I think this might be overkill. What the user needs to know is that the tree wasn't finalized, which they learn from the error return. If they want to know the specific length they can call `.len()` themselves.

Reply:
There's also possibly nicer error message. But if people find it too much effort, it's OK.